ethereum/blockwatch/block_watcher.go

```diff
@@ -270,3 +289,257 @@ func (w *Watcher) addLogs(header *meshdb.MiniHeader) (*meshdb.MiniHeader, error)
+			return events, nil
+		}
+
+		// Create the block events from all the logs found by grouping
+		// them into blockHeaders
+		hashToBlockHeader := map[common.Hash]*meshdb.MiniHeader{}
+		for _, log := range logs {
+			blockHeader, ok := hashToBlockHeader[log.BlockHash]
+			if !ok {
+				// TODO(fabio): Find a way to include the parent hash for the block as well.
+				// It's currently not an issue to omit it since we don't use the parent hash
+				// when processing block events in OrderWatcher.
+				blockHeader = &meshdb.MiniHeader{
+					Hash:   log.BlockHash,
+					Number: big.NewInt(0).SetUint64(log.BlockNumber),
+					Logs:   []types.Log{},
+				}
+				hashToBlockHeader[log.BlockHash] = blockHeader
+			}
+			blockHeader.Logs = append(blockHeader.Logs, log)
+		}
+		for _, blockHeader := range hashToBlockHeader {
+			events = append(events, &Event{
+				Type:        Added,
+				BlockHeader: blockHeader,
+			})
+		}
+		return events, nil
+	}
+	return events, nil
+}
+
+type logRequestResult struct {
+	From                   int64
+	To                     int64
+	Logs                   []types.Log
+	Err                    error
+	FurthestBlockProcessed int64
+}
+
+// getLogsInBlockRange attempts to fetch all logs in the block range specified. If it retrieves
+// all logs in the range, it returns them. If it fails to retrieve some of the blocks, it
+// returns all the logs it did find, along with the block number after which no further logs
+// were retrieved.
+func (w *Watcher) getLogsInBlockRange(from, to int64) ([]types.Log, int64) {
+	chunks := w.getBlockRangeChunks(from, to, maxBlocksInGetLogsQuery)
+
+	mu := sync.Mutex{}
+	indexToLogResult := map[int]logRequestResult{}
+	didARequestFail := false
+
+	semaphoreChan := make(chan struct{}, concurrencyLimit)
+	defer close(semaphoreChan)
+
+	wg := &sync.WaitGroup{}
+	for i, chunk := range chunks {
+		wg.Add(1)
+		go func(index int, chunk *blockRange) {
+			defer wg.Done()
+
+			// Add one to the semaphore chan. If it already has concurrencyLimit values,
+			// the request blocks here until one frees up.
+			semaphoreChan <- struct{}{}
+			// If a previous request failed, we noop all subsequent requests
+			mu.Lock()
+			if didARequestFail {
+				mu.Unlock()
+				<-semaphoreChan
+				return
+			}
+			mu.Unlock()
+
+			logs, err := w.filterLogsRecurisively(chunk.FromBlock, chunk.ToBlock, []types.Log{})
+			if err != nil {
+				log.WithField("error", err).Trace("Failed to fetch logs for ", chunk.FromBlock, " to ", chunk.ToBlock)
+				mu.Lock()
+				didARequestFail = true
+				mu.Unlock()
+			}
+			mu.Lock()
+			indexToLogResult[index] = logRequestResult{
+				From: chunk.FromBlock,
+				To:   chunk.ToBlock,
+				Err:  err,
+				Logs: logs,
+			}
+			mu.Unlock()
+			<-semaphoreChan
+		}(i, chunk)
+	}
+
+	// Wait for all log requests to complete
+	wg.Wait()
+
+	logs := []types.Log{}
+	for i := range chunks {
+		logRequestResult := indexToLogResult[i]
+		if logRequestResult.Err != nil {
```

Comment:
Yes, you are correct. `i` might fail, and `i+1` might succeed. We intentionally discard the results in `i+1` because we need the contiguous history of events and cannot be missing gaps. Because of this, we only process up until the first error (if one exists), and discard any responses returned beyond that point.